client/consensus/grandpa/authorities.go

```diff
@@ -184,27 +273,360 @@ func (authSet *AuthoritySet) addStandardChange(pending PendingChange, isDescende
+// authority set change (without triggering it), ensuring that if there are
+// multiple changes in the same branch, finalizing this block won't
+// finalize past multiple transitions (i.e. transitions must be finalized
+// in-order). Returns *true if the block being finalized enacts a
+// change that can be immediately applied, *false if the block being
+// finalized enacts a change but it cannot be applied yet since there are
+// other dependent changes, and nil if no change is enacted. The given
+// function `is_descendent_of` should return `true` if the second hash
+// (target) is a descendent of the first hash (base).
+func (authSet *AuthoritySet) EnactsStandardChange(
+	finalizedHash common.Hash, finalizedNumber uint, isDescendentOf IsDescendentOf) (*bool, error) {
+	applied, err := authSet.pendingStandardChanges.FinalizeAnyWithDescendentIf(&finalizedHash, finalizedNumber, isDescendentOf, enactStandardChangesPredicate(finalizedNumber))
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", errForkTree, err)
+	}
+	return applied, nil
 }
 
-// InnerLocked
-// Returns access to the [`AuthoritySet`] and locks it.
+// PendingChange A pending change to the authority set.
 //
-// For more information see [`SharedDataLocked`].
-// TODO update with shared logic
-func (sas *SharedAuthoritySet) InnerLocked() AuthoritySet {
-	return sas.authoritySet
+// This will be applied when the announcing block is at some depth within
+// the finalized or unfinalized chain.
+type PendingChange struct {
+	// The new authorities and weights to apply.
+	nextAuthorities AuthorityList
+	// How deep in the chain the announcing block must be
+	// before the change is applied.
+	delay uint
+	// The announcing block's height.
+	canonHeight uint
+	// The announcing block's hash.
+	canonHash common.Hash
+	// The delay kind.
+	delayKind DelayKind
+}
+
+// EffectiveNumber Returns the effective number this change will be applied at.
+func (pc *PendingChange) EffectiveNumber() uint {
+	return pc.canonHeight + pc.delay
+}
+
+// AuthoritySetChanges Tracks historical authority set changes. We store the block numbers for the last block
+// of each authority set, once they have been finalized. These blocks are guaranteed to
+// have a justification unless they were triggered by a forced change.
+type AuthoritySetChanges []authorityChange
+
+// append an authorityChange to AuthoritySetChanges
+func (authSetChanges *AuthoritySetChanges) append(setId uint64, blockNumber uint) {
+	*authSetChanges = append(*authSetChanges, authorityChange{
+		setId:       setId,
+		blockNumber: blockNumber,
+	})
+}
+
+// Three states that can be returned: Latest, Set (tuple), Unknown
+// Latest => bool
+// Set => &AuthorityChange
+// Unknown => nil
+// TODO for reviewers, this can be a VDT but I'm not sure its needed
+func (authSetChanges *AuthoritySetChanges) getSetId(blockNumber uint) (latest bool, set *authorityChange, err error) {
```

Comment:
It can go either way. Can use this if we don't want the overhead of a vdt type, but also would work well with vdt too. Can update if you think it is best